src/node/rpc/jwt_management.h

```diff
@@ -192,38 +243,132 @@ namespace ccf
       }
       LOG_INFO_FMT("{}: Storing JWT signing key with kid {}", log_prefix, kid);
       new_keys.emplace(kid, der);
+
+      if (jwk.issuer)
+      {
+        if (!check_issuer(issuer, *jwk.issuer))
+        {
+          LOG_FAIL_FMT(
+            "{}: JWKS kid {} with issuer constraint {} fails validation "
+            "against issuer {}",
+            log_prefix,
+            kid,
+            *jwk.issuer,
+            issuer);
+          return false;
+        }
+
+        issuer_constraints.emplace(kid, *jwk.issuer);
+      }
     }
+
     if (new_keys.empty())
     {
       LOG_FAIL_FMT("{}: no keys left after applying filter", log_prefix);
       return false;
     }
 
     std::set<std::string> existing_kids;
-    key_issuer->foreach(
-      [&existing_kids, &issuer](const auto& kid, const auto& issuer_) {
-        if (issuer_ == issuer)
+    std::set<std::string> kids_with_new_constraints;
+    key_issuers->foreach([&existing_kids,
+                          &issuer_constraints,
+                          &kids_with_new_constraints,
+                          &issuer](const auto& kid, const auto& issuers) {
+      for (const auto& issuer_with_constraint : issuers)
+      {
+        if (issuer == issuer_with_constraint.issuer)
         {
           existing_kids.insert(kid);
+
+          const auto it = issuer_constraints.find(kid);
+          if (
+            it != issuer_constraints.end() &&
+            it->second == issuer_with_constraint.constraint)
+          {
+            kids_with_new_constraints.insert(kid);
+          }
+
+          break; // 1 issuer <-> 1 kid
         }
-        return true;
-      });
+      }
+      return true;
+    });
 
     for (auto& [kid, der] : new_keys)
     {
-      if (!existing_kids.contains(kid))
+      const bool new_kid = !existing_kids.contains(kid);
+      const bool new_constraint = !kids_with_new_constraints.contains(kid);
+
+      if (new_kid)
       {
         keys->put(kid, der);
-        key_issuer->put(kid, issuer);
+      }
+
+      if (new_kid || new_constraint)
+      {
+        JwtIssuerWithConstraint value{issuer, std::nullopt};
+        const auto it = issuer_constraints.find(kid);
+        if (it != issuer_constraints.end())
+        {
+          value.constraint = it->second;
+        }
+
+        LOG_INFO_FMT(
```

Comment:
```suggestion
        LOG_DEBUG_FMT(
```

I think `INFO` is too verbose for this, and will lead to a lot of logging spam in the output.